### PR TITLE
GH-3135 : Adding new `ListenerContainerRegistry.getListenerContainersMatchingId(Predicate)` for easier usage. 

### LINKED ISSUE #3135: Adding new `KafkaListenerEndpointRegistry.getListenerContainersMatchingId(predicate: Predicate<String>)`
**Expected Behavior**
Would be nice to have method...

```java
kafkaListenerEndpointRegistry
            .getListenerContainersMatchingId( id -> id.startsWith(prefix))
           .forEach(MessageListenerContainer::start)
           
```

**Current Behavior**

```java
kafkaListenerEndpointRegistry
            .getListenerContainerIds()
            .filter( id -> id.startsWith(prefix))
            .forEach(id -> kafkaListenerEndpointRegistry.getListenerContainer(id).start())
```

**Context**

I run application that starts/stops listeners at certain time(office hours)

We have listeners with ids such as...

- `listener_product_purchase`
- `listener_product_purchase-retry-0`
- `listener_product_purchase-retry-1`
- `listener_product_purchase-dlt`

... this, generated by @RetryableTopic configuration.
Just some syntactic sugar.

### PR DESCRIPTION
**Pull request by JooHyukKim**
Resolves #3135

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/config/KafkaListenerEndpointRegistry.java [hunk 1]
```diff
@@ -115,6 +118,25 @@ public MessageListenerContainer getListenerContainer(String id) {
...
+	 * @see #getListenerContainer(String)
+	 */
+	@Override
+	public Collection<MessageListenerContainer> getListenerContainersMatching(Predicate<String> idMatcher) {
+		Assert.notNull(idMatcher, "'idMatcher' cannot be null");
+		List<MessageListenerContainer> matchingContainers = this.listenerContainers.entrySet()
+			.stream()
+			.filter(entry -> idMatcher.test(entry.getKey()))
+			.map(Map.Entry::getValue)
+			.collect(Collectors.toList());
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/KafkaListenerEndpointRegistry.java:136]: There is just `toList()` API.

File: spring-kafka/src/main/java/org/springframework/kafka/config/KafkaListenerEndpointRegistry.java [hunk 2]
```diff
@@ -115,6 +118,25 @@ public MessageListenerContainer getListenerContainer(String id) {
 		return this.listenerContainers.get(id);
 	}
 
+	/**
+	 * Return all {@link MessageListenerContainer} instances with id matching the predicate or
+	 * empty {@link Collection} if no such container exists.
+	 * @param idMatcher the predicate to match container id with
+	 * @return the containers or empty {@link Collection} if no container with that id exists
+	 * @see #getListenerContainerIds()
+	 * @see #getListenerContainer(String)
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/KafkaListenerEndpointRegistry.java:127]: `@since 3.2`

File: spring-kafka/src/main/java/org/springframework/kafka/listener/ListenerContainerRegistry.java [hunk 3]
```diff
@@ -41,6 +42,16 @@ public interface ListenerContainerRegistry {
...
 	MessageListenerContainer getListenerContainer(String id);
 
+	/**
+	 * Return all {@link MessageListenerContainer} instances with id matching the predicate or
+	 * empty {@link Collection} if no such container exists.
+	 * @param idMatcher the predicate to match the container id with
+	 * @return the containers or empty {@link Collection} if no container with that id exists
+	 * @see #getListenerContainerIds()
+	 * @see #getListenerContainer(String)
+	 */
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/ListenerContainerRegistry.java:52]: `@since 3.2`

File: spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java [hunk 4]
```diff
@@ -47,4 +63,77 @@ void unregister() {
...
+	@DisplayName("getListenerContainersMatching should return unmodifiable list")
+	@Test
+    void testGetListenerContainersMatchingReturnsUnmodifiableList() {
+        // Given
+        KafkaListenerEndpointRegistry registry = new KafkaListenerEndpointRegistry();
+        registerListenerWithId(registry, "foo");
+        // When
+        Collection<MessageListenerContainer> listeners = registry.getListenerContainersMatching(s -> true);
+        // Then
+        assertThrows(UnsupportedOperationException.class, () -> listeners.add(mock(MessageListenerContainer.class)));
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java:87]: Why don't use `assertThatExceptionOfType(UnsupportedOperationException.class)` instead?

It also probably going to fail on Checkstyle since we have a limited set of allowed `static` method imports.

File: spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java [hunk 5]
```diff
@@ -47,4 +63,77 @@ void unregister() {
...
+			// Corner case: No names match as the predicate always returns false
+			Arguments.of(names, (Predicate<String>) id -> false, 0)
+		);
+	}
+
+	private void registerWithListenerIds(KafkaListenerEndpointRegistry registry, List<String> names) {
+		names.forEach(name -> registerListenerWithId(registry, name));
+	}
+
+	private void registerListenerWithId(KafkaListenerEndpointRegistry registry, String id) {
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java:130]: Looks like these both methods can be `static`.

**Review by artembilan — CHANGES_REQUESTED** [review]
Please, add your name to the `@author` list of all the affected classes.

Also consider to mention this new feature in the `whats-new.adoc` and in the `kafkalistener-lifecycle.adoc`.

Thanks

File: spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java (hunk 4 above)
Comment (JooHyukKim) [reply spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java]: Right, looking now there is no pre-existing usage of `assertThatExceptionOfType()`

**Comment by JooHyukKim** [comment]
I've updated the author list and added documentation for the new feature in `whats-new.adoc` and `kafkalistener-lifecycle.adoc`.

I noticed there seems to be limited documentation on `KafkaListenerEndpointRegistry`, particularly methods like `getListenerContainers()` and `getListenerContainer(String)`. I can
 address this in a future PR to enhance docs further. WDYT, @artembilan?

**Comment by artembilan** [comment]
Sure! You can add those couple sentences just in this PR. We have a release on Monday, so it will be reviewed immediately .
Thanks

**Comment by JooHyukKim** [comment]
Added some more doc as discussed 👍🏼 cc. @artembilan

File: spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc [hunk 6]
```diff
@@ -83,3 +83,13 @@ See xref:kafka/seek.adoc#seek[Seek API Docs] for more details.
 For Spring Boot applications which define an application name, this name is now used
 as a default prefix for auto-generated client IDs for certain client types.
 See xref:kafka/connecting.adoc#default-client-id-prefixes[Default client ID prefixes] for more details.
+
+[[get-listener-containers-matching]]
+== Retrieve MessageListenerContainers with by matching a Predicate
+
+Starting with version 3.2, the `ListenerContainerRegistry` introduces the `getListenerContainersMatching` method.
+This method filters `MessageListenerContainer` instances based on a `Predicate<String>` applied to their IDs.
```
Comment (artembilan) [inline spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc:91]: After your great docs I wonder if we could extend this new feature to something like this `BiPredicate<String, MessageListenerContainer>`. This way we may filter desired containers not only by their ids, but via some other option form the container itself.

WDYT?

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc (hunk 6 above)
Comment (JooHyukKim) [reply spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc:91]: Makes sense 👍🏼
I wonder if we should have both or just `BiPredicate<String, MessageListenerContainer>` 🤔
Extended feature sounds very powerful, but might be too verbose in someway, knowing that `id` is the key used to access the `MessageContainerListener`.

Comment (JooHyukKim) [reply spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc:91]: Since the `String, MessageListenerContainer` pair in `BiPredicate` technically represents the key-value pair stored in the registry, how about having both?

Comment (JooHyukKim) [reply spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc:91]: Should we do a separate PR or do everything in this one?

Comment (artembilan) [reply spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc:91]: This one is OK.
Thanks

Comment (JooHyukKim) [reply spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc:91]: Alright, hopefully it would probably take an hour or two

File: spring-kafka/src/main/java/org/springframework/kafka/listener/ListenerContainerRegistry.java [hunk 7]
```diff
@@ -41,6 +44,29 @@ public interface ListenerContainerRegistry {
...
+	 * Return all {@link MessageListenerContainer} instances that satisfy the given bi-predicate.
+	 * The {@code BiPredicate<String, MessageListenerContainer>} takes the container id and the container itself as arguments.
+	 * This allows for more sophisticated filtering, including properties or state of the container itself.
+	 * @param idAndContainerMatcher the bi-predicate to match the container id and the container
+	 * @return the containers that match the bi-predicate criteria or an empty {@link Collection} if no matching containers exist
+	 * @since 3.2
+	 * @see #getListenerContainerIds()
+	 * @see #getListenerContainersMatching(Predicate)
+	 */
+	Collection<MessageListenerContainer> getListenerContainersMatching(BiPredicate<String, MessageListenerContainer> idAndContainerMatcher);
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/ListenerContainerRegistry.java:68]: Let's see if you can follow the rule no more than 120 symbols per line!

File: spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc [hunk 8]
```diff
@@ -83,3 +83,26 @@ See xref:kafka/seek.adoc#seek[Seek API Docs] for more details.
 For Spring Boot applications which define an application name, this name is now used
 as a default prefix for auto-generated client IDs for certain client types.
 See xref:kafka/connecting.adoc#default-client-id-prefixes[Default client ID prefixes] for more details.
+
+[[get-listener-containers-matching]]
+== Enhanced Retrieval of MessageListenerContainers
+
+Introduced in version 3.2, the `ListenerContainerRegistry` provides advanced methods for dynamically filtering `MessageListenerContainer` instances, enhancing management capabilities.
```
Comment (artembilan) [inline spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc:90]: This file is fully specific to particular version we mention in the beginning of the file.
More over the content of the file is just a list of topics with links to details.
You do show enough in the `kafkalistener-lifecycle.adoc`, but that all must not be duplicated in `What's New` file.

File: spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java [hunk 9]
```diff
@@ -47,4 +60,123 @@ void unregister() {
...
+	private static void registerListenerWithId(KafkaListenerEndpointRegistry registry, String id) {
+		KafkaListenerEndpoint endpoint = mock(KafkaListenerEndpoint.class);
+		@SuppressWarnings("unchecked")
+		KafkaListenerContainerFactory<MessageListenerContainer> factory = mock(KafkaListenerContainerFactory.class);
+		given(endpoint.getId()).willReturn(id);
+		MessageListenerContainer container = mock(MessageListenerContainer.class);
+		given(container.isRunning()).willReturn(true);
+		given(factory.createListenerContainer(endpoint)).willReturn(container);
+		registry.registerListenerContainer(endpoint, factory);
+	}
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java:181]: Every member (including the last method of the last class) in the class must be surrounded with blank lines for much better code readability.

**Review by artembilan — CHANGES_REQUESTED** [review]
The API is great: only some cosmetic concerns in regards of our code style rules.
Thanks

File: spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc (hunk 8 above)
Comment (JooHyukKim) [reply spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc]: True, thank you for pointing out!
